jedi/evaluate/pep0484.py

```diff
@@ -60,3 +81,81 @@ def follow_param(evaluator, param):
 def find_return_types(evaluator, func):
     annotation = func.py__annotations__().get("return", None)
     return _evaluate_for_annotation(evaluator, annotation)
+
+
+# TODO: Memoize
+def _get_typing_replacement_module():
+    """
+    The idea is to return our jedi replacement for the PEP-0484 typing module
+    as discussed at https://github.com/davidhalter/jedi/issues/663
+    """
+
+    typing_path = os.path.abspath(os.path.join(__file__, "../jedi_typing.py"))
+    with open(typing_path) as f:
+        code = f.read()
+    p = ParserWithRecovery(load_grammar(), code)
+    return p.module
+
+
+def get_types_for_typing_module(evaluator, typ, node):
+    from jedi.evaluate.iterable import FakeSequence
+    if not typ.base.get_parent_until().name.value == "typing":
+        return None
+    # we assume that any class using [] in a module called
+    # "typing" with a name for which we have a replacement
+    # should be replaced by that class. This is not 100%
+    # airtight but I don't have a better idea to check that it's
+    # actually the PEP-0484 typing module and not some other
+    if tree.is_node(node, "subscriptlist"):
+        nodes = node.children[::2]  # skip the commas
+    else:
+        nodes = [node]
+    del node
+
+    # hacked in Union and Optional, since it's hard to do nicely in parsed code
+    if typ.name.value == "Union":
+        return set().union(*[evaluator.eval_element(node) for node in nodes])
```

Comment:
I think `iterable.unite` would be pretty nice for this. I'm using it quite often. I know its name kind might be confusing, if you have a name in mind, please tell me.